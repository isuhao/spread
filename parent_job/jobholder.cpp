#include "jobholder.hpp"
#include <assert.h>
#include <job/thread.hpp>

#define LOCK boost::lock_guard<boost::mutex> lock(mutex)

using namespace Spread;

void JobHolder::clearFailed()
{
  LOCK;
  done.clear();
}

void JobHolder::doJob()
{
  startup();
  start();

  while(true)
    {
      /* Loop through jobs. We use special functions to ensure
         thread safety, since new jobs may be inserted at any time.
      */
      bool empty = false;
      {
        LOCK;
        JobSet::const_iterator it, it2;
        for(it = jobs.begin(); it != jobs.end();)
          {
            it2 = it++;

            // Simply remove all finished jobs
            if((*it2)->getInfo()->isFinished())
              {
                JobPtr p = *it2;
                jobs.erase(it2);

                // Keep failed jobs to help the user debug
                if(p->getInfo()->isError())
                  {
                    handleError(p->getInfo()->getMessage());
                    done.push_back(p);
                  }
              }
          }

        empty = jobs.size() == 0;
        if(finishOnEmpty && empty)
          setDone();
      }

      // This will exit if setDone() was called, or if the job was
      // aborted.
      if(checkStatus()) return;

      tick();

      Thread::sleep(0.1);

      // Sleep some more if there is nothing to do
      if(empty)
        Thread::sleep(0.5);
    }
}
