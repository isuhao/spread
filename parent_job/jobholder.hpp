#ifndef __SPREAD_JOBHOLDER_HPP_
#define __SPREAD_JOBHOLDER_HPP_

#include "listjob.hpp"

/* JobHolder is a ListJob that holds a list of independently running
   thread jobs. New jobs are added with the add() functions (from
   ListJob). It is useful as a "top-level" interface object between a
   tree of running tasks, and the main application thread.

   Jobs are removed from the list when they succeed or are
   aborted. However failed jobs are kept in the 'done' list so the
   user can debug them.

   The JobHolder will instantly abort all its children if it is itself
   aborted.
*/

namespace Spread
{
  struct JobHolder : ListJob
  {
    // Called right after doJob() starts
    virtual void startup() {}

    // Called regularly by doJob() while the task runs
    virtual void tick() {}

    // When set to true, the job will finish when it runs out of jobs.
    bool finishOnEmpty;

    // Tell the job to terminate when all jobs have ended.
    void finish() { finishOnEmpty = true; }

    // Wait for all jobs to terminate
    void waitFinish() { finish(); info->wait(); }

    JobHolder() : finishOnEmpty(false) {}

  private:
    void doJob();
  };
}
#endif
