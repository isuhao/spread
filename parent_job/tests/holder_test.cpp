#include "jobholder.hpp"
#include "common.cpp"

struct TestJob : Job
{
  int i;
  TestJob(int i_) : i(i_) {}
  ~TestJob() { cout << "Deleting i=" << i << endl; }

  void doJob()
  {
    cout << "Running i=" << i << endl;

    if(i == 2) Thread::sleep(3);
    if(i == 3) throw runtime_error("Failing yo!");
    setDone();
  }
};

struct MyJobHolder : JobHolder
{
  void startup() { cout << "STARTUP\n"; }
  void tick() { cout << "TICK\n"; }
};

int main()
{
  JobHolder *jh = new MyJobHolder;
  JobPtr ptr(jh);

  Thread::run(ptr);
  print(ptr);
  cout << "Adding jobs:\n";
  jh->add(new TestJob(1));
  jh->add(new TestJob(2));
  jh->add(new TestJob(3));
  print(ptr);
  cout << "\nWaiting for finish:\n";
  jh->waitFinish();
  print(ptr);

  cout << "\nClearing failed list:\n";
  jh->clearFailed();
  print(ptr);

  cout << "EXIT\n";
  return 0;
}
