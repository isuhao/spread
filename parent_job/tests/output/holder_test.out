UNK:  0/0
Adding jobs:
UNK:  0/0
  JOBS:
  UNK:  0/0
  UNK:  0/0
  UNK:  0/0

Waiting for finish:
STARTUP
TICK
Running i=3
Running i=2
Running i=1
Deleting i=1
TICK
TICK
TICK
TICK
TICK
Deleting i=2
SUC: Success 0/0
  DONE:
  ERR: Failing yo! 0/0

Clearing failed list:
Deleting i=3
SUC: Success 0/0
EXIT
